balancer-js/src/modules/pools/pool-types/concerns/composableStable/join.concern.ts

```diff
@@ -6,115 +6,358 @@ import {
-    const expectedBPTOut = StableMathBigInt._calcBptOutGivenExactTokensIn(
-      BigInt(parsedAmp),
-      parsedBalancesWithoutBpt.map(BigInt), // Should not have BPT
-      scaledAmountsIn, // Should not have BPT
-      BigInt(parsedTotalShares),
-      BigInt(parsedSwapFee)
-    ).toString();
+  /**
+   * Sorts inputs and pool value to be correct order and scale for maths and Vault interaction.
+   * @param values
+   * @returns
+   */
+  sortValuesBasedOnPoolVersion(
+    values: Pick<
+      JoinPoolParameters,
+      'pool' | 'wrappedNativeAsset' | 'amountsIn' | 'tokensIn'
+    >
+  ): SortedValues {
+    /**
+     * V1: Does not have proportional exits.
+     * V2: Reintroduced proportional exits. Has vulnerability.
+     * V3: Fixed vulnerability. Functionally the same as V2.
+     */
+    if (values.pool.poolTypeVersion < 4)
+      return this.sortV1(
+        values.wrappedNativeAsset,
+        values.tokensIn,
+        values.amountsIn,
+        values.pool
+      );
+    // Not release yet and needs tests to confirm
+    // else if (values.pool.poolTypeVersion === 4)
+    //   sortedValues = this.sortV4(
+    //     values.tokensIn,
+    //     values.amountsIn,
+    //     values.pool
+    //   );
+    else
+      throw new Error(
+        `Unsupported ComposablePool Version ${values.pool.poolTypeVersion}`
+      );
+  }
+
+  /**
+   * Ensure tokensIn and amountsIn match pool tokens length
+   * @param tokensIn
+   * @param amountsIn
+   * @param poolTokens
+   */
+  checkInputs(
+    tokensIn: string[],
+    amountsIn: string[],
+    poolTokens: string[]
+  ): void {
+    if (
+      tokensIn.length != amountsIn.length ||
+      tokensIn.length != poolTokens.length - 1
+    ) {
+      throw new BalancerError(BalancerErrorCode.INPUT_LENGTH_MISMATCH);
+    }
+  }
 
+  /**
+   * Encodes user data with slippage applied to expected BPT out.
+   * @param expectedBPTOut
+   * @param slippage
+   * @param amountsIn
+   * @returns
+   */
+  encodeUserDataExactTokensInForBPTOut(
+    expectedBPTOut: bigint,
+    slippage: string,
+    amountsIn: string[]
+  ): { userData: string; minBPTOut: string } {
     const minBPTOut = subSlippage(
       BigNumber.from(expectedBPTOut),
       BigNumber.from(slippage)
     ).toString();
 
     //NEEDS TO ENCODE USER DATA WITHOUT BPT AMOUNT
-    const userData = ComposableStablePoolEncoder.joinExactTokensInForBPTOut(
-      sortedAmountsIn,
-      minBPTOut
-    );
+    return {
+      userData: ComposableStablePoolEncoder.joinExactTokensInForBPTOut(
+        amountsIn, // No BPT amount
+        minBPTOut
+      ),
+      minBPTOut,
+    };
+  }
 
+  /**
+   * Encode transaction data into an ABI byte string which can be sent to the network to be executed
+   * @param poolId
+   * @param sender
+   * @param recipient
+   * @param assetsWithBpt
+   * @param encodedUserData
+   * @param maxAmountsInWithBpt
+   * @returns
+   */
+  encodeJoinPool(
```

Comment:
It feels a bit odd that `encodeJoinPool` is returning both encoded and non-encoded data. Definitely not critical if we want to keep it like this.

Reply:
Not sure of the best alternative?

Reply:
We could break it down into smaller functions, but I'm not sure it would be indeed a better alternative. I'd say it's fine as is 👍